File: java/flowcamel-core/src/main/java/com/flowcamel/core/graph/GraphValidator.java
package com.flowcamel.core.graph;

import com.flowcamel.core.config.ApplicationConfigEmitter;
import com.flowcamel.core.config.ConfigRefs;
import com.flowcamel.core.model.BlockCategory;
import com.flowcamel.core.model.BlockDefinition;
import com.flowcamel.core.model.FlowDefinition;
import com.flowcamel.core.model.FlowGraph;
import com.flowcamel.core.model.FlowNode;
import com.flowcamel.core.model.PropSchema;
import com.flowcamel.core.model.ValidationResult;
import com.flowcamel.core.properties.ComponentProperties;
import com.flowcamel.core.registry.BlockRegistry;
import com.flowcamel.core.uri.UriBuilder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class GraphValidator {
  private GraphValidator() {}

  public static ValidationResult validateForYamlExport(FlowGraph graph) {
    graph = GraphNormalizer.normalize(graph);
    List<String> errors = new ArrayList<>();
    List<FlowDefinition> flows = graph.flows;

    Map<String, Integer> routeCounts = new HashMap<>();
    Set<String> routeIds = new HashSet<>();
    for (FlowDefinition f : flows) {
      if (f.routeId != null && !f.routeId.isBlank()) {
        routeCounts.merge(f.routeId.trim(), 1, Integer::sum);
        routeIds.add(f.routeId.trim());
      }
    }
    for (var e : routeCounts.entrySet()) {
      if (e.getValue() > 1) errors.add("Duplicate route id \"" + e.getKey() + "\" across flows.");
    }

    boolean anyNodes = false;
    for (FlowDefinition flow : flows) {
      if (flow.nodes == null || flow.nodes.isEmpty()) continue;
      anyNodes = true;
      errors.addAll(validateFlowForYaml(flow, graph, routeIds));
    }
    if (!anyNodes) errors.add("Project has no blocks in any flow.");

    return new ValidationResult(errors.isEmpty(), errors);
  }

  public static ValidationResult validate(FlowGraph graph) {
    graph = GraphNormalizer.normalize(graph);
    List<String> errors = new ArrayList<>();
    Set<String> routeIds = new HashSet<>();
    for (FlowDefinition f : graph.flows) {
      if (f.routeId != null) routeIds.add(f.routeId.trim());
    }
    boolean anyNodes = false;
    for (FlowDefinition flow : graph.flows) {
      if (flow.nodes == null || flow.nodes.isEmpty()) continue;
      anyNodes = true;
      errors.addAll(validateFlow(flow, routeIds));
    }
    if (!anyNodes) errors.add("Add at least one block to a flow.");
    return new ValidationResult(errors.isEmpty(), errors);
  }

  private static List<String> validateFlow(FlowDefinition flow, Set<String> routeIds) {
    List<String> errors = new ArrayList<>();
    String prefix = "[" + flow.name + "] ";

    long sources =
        flow.nodes.stream()
            .filter(n -> BlockRegistry.getBlock(n.blockType).map(b -> b.category == BlockCategory.SOURCE).orElse(false))
            .count();
    boolean hasCallFlow = flow.nodes.stream().anyMatch(n -> "call-flow-action".equals(n.blockType));
    long dests =
        flow.nodes.stream()
            .filter(
                n ->
                    BlockRegistry.getBlock(n.blockType)
                        .map(b -> b.category == BlockCategory.DESTINATION)
                        .orElse(false))
            .count();

    if (!flow.nodes.isEmpty() && sources == 0) {
      errors.add(prefix + "Flow must have at least one source block.");
    }
    if (!flow.nodes.isEmpty() && dests == 0 && !hasCallFlow) {
      errors.add(prefix + "Flow must have a destination or a Call flow block.");
    }

    Map<String, Integer> incoming = new HashMap<>();
    Map<String, Integer> outgoing = new HashMap<>();
    for (FlowNode n : flow.nodes) {
      incoming.put(n.id, 0);
      outgoing.put(n.id, 0);
    }
    for (var e : flow.edges) {
      outgoing.merge(e.source, 1, Integer::sum);
      incoming.merge(e.target, 1, Integer::sum);
    }

    for (FlowNode node : flow.nodes) {
      BlockCategory cat = BlockRegistry.getBlock(node.blockType).map(b -> b.category).orElse(null);
      int inc = incoming.getOrDefault(node.id, 0);
      int out = outgoing.getOrDefault(node.id, 0);
      if (cat == BlockCategory.SOURCE && inc > 0) {
        errors.add(prefix + "Source block \"" + node.label + "\" must not have incoming connections.");
      }
      if (cat == BlockCategory.DESTINATION && out > 0) {
        errors.add(prefix + "Destination block \"" + node.label + "\" must not have outgoing connections.");
      }
      if (cat != BlockCategory.SOURCE && inc == 0) {
        errors.add(prefix + "Block \"" + node.label + "\" has no incoming connection.");
      }
      if (cat != BlockCategory.DESTINATION && cat != BlockCategory.ACTION && out == 0) {
        errors.add(prefix + "Block \"" + node.label + "\" has no outgoing connection.");
      }
      if ("call-flow-action".equals(node.blockType) && out > 0) {
        errors.add(prefix + "\"" + node.label + "\" (Call flow) must not have outgoing connections.");
      }
    }
    if (hasCycle(flow)) errors.add(prefix + "Flow contains a cycle, which is not allowed.");
    return errors;
  }

  private static List<String> validateFlowForYaml(
      FlowDefinition flow, FlowGraph graph, Set<String> routeIds) {
    List<String> errors = new ArrayList<>(validateFlow(flow, routeIds));
    String prefix = "[" + flow.name + "] ";

    for (FlowNode node : flow.nodes) {
      BlockDefinition block = BlockRegistry.getBlock(node.blockType).orElse(null);
      if (block == null) {
        errors.add(prefix + "Unknown block type \"" + node.blockType + "\" on \"" + node.label + "\".");
        continue;
      }
      Map<String, String> props = ComponentProperties.resolveNodeProps(node.blockType, node.props);
      for (PropSchema schema : ComponentProperties.getWizardSteps(node.blockType)) {
        if (!schema.required) continue;
        if (ComponentProperties.resolvePropValue(props, schema).isEmpty()) {
          errors.add(
              prefix
                  + "\""
                  + node.label
                  + "\" ("
                  + block.label
                  + "): missing \""
                  + (schema.label != null ? schema.label : schema.key)
                  + "\".");
        }
      }
      if ("call-flow-action".equals(node.blockType)) {
        String target = props.getOrDefault("targetRouteId", "").trim();
        if (target.isEmpty()) {
          errors.add(prefix + "\"" + node.label + "\" (Call flow): select a target flow.");
        } else if (target.equals(flow.routeId)) {
          errors.add(prefix + "\"" + node.label + "\" (Call flow): cannot call the same flow.");
        } else if (!routeIds.contains(target)) {
          errors.add(prefix + "\"" + node.label + "\" (Call flow): unknown route \"" + target + "\".");
        }
      }
      boolean needsUri =
          block.category == BlockCategory.SOURCE
              || block.category == BlockCategory.DESTINATION
              || (block.camelUri != null && !block.camelUri.isEmpty());
      if (needsUri
          && !"call-flow-action".equals(node.blockType)
          && UriBuilder.buildEndpointUri(node.blockType, props).isEmpty()) {
        errors.add(
            prefix
                + "\""
                + node.label
                + "\" ("
                + block.label
                + "): endpoint URI could not be built — check required settings.");
      }
      if (block.category == BlockCategory.ACTION) {
        String expression = props.get("expression");
        if ("filter-action".equals(node.blockType)
            && (expression == null || expression.trim().isEmpty())
            && !ConfigRefs.isConfigRef(expression)) {
          errors.add(prefix + "\"" + node.label + "\" (Filter): missing filter expression.");
        }
        if ("transform-action".equals(node.blockType)
            && (expression == null || expression.trim().isEmpty())
            && !ConfigRefs.isConfigRef(expression)) {
          errors.add(prefix + "\"" + node.label + "\" (Transform): missing expression.");
        }
        if (node.props != null) {
          for (var e : node.props.entrySet()) {
            if (ConfigRefs.isConfigRef(e.getValue())) {
              String cfgKey = ConfigRefs.configRefKey(e.getValue());
              if (!ApplicationConfigEmitter.configKeyExists(graph.config, cfgKey)) {
                errors.add(
                    prefix
                        + "\""
                        + node.label
                        + "\": property \""
                        + e.getKey()
                        + "\" references unknown config key \""
                        + cfgKey
                        + "\".");
              }
            }
          }
        }
      }
    }
    return errors;
  }

  private static boolean hasCycle(FlowDefinition flow) {
    Map<String, List<String>> adj = new HashMap<>();
    for (FlowNode n : flow.nodes) adj.put(n.id, new ArrayList<>());
    for (var e : flow.edges) adj.computeIfAbsent(e.source, k -> new ArrayList<>()).add(e.target);

    Map<String, Integer> color = new HashMap<>();
    for (FlowNode n : flow.nodes) color.put(n.id, 0);

    for (FlowNode n : flow.nodes) {
      if (color.get(n.id) == 0 && dfsCycle(n.id, adj, color)) return true;
    }
    return false;
  }

  private static boolean dfsCycle(String id, Map<String, List<String>> adj, Map<String, Integer> color) {
    color.put(id, 1);
    for (String neighbor : adj.getOrDefault(id, List.of())) {
      if (color.get(neighbor) == 1) return true;
      if (color.get(neighbor) == 0 && dfsCycle(neighbor, adj, color)) return true;
    }
    color.put(id, 2);
    return false;
  }
}

File: packages/core/src/api/GraphValidator.ts
import { BlockCategory } from '../model/BlockCategory.js';
import type { FlowDefinition } from '../model/FlowDefinition.js';
import { FlowGraph } from '../model/FlowGraph.js';
import { getBlock } from './BlockRegistry.js';
import { getWizardSteps, resolveNodeProps, resolvePropValue } from './ComponentProperties.js';
import { configKeyExists } from './ApplicationConfigYaml.js';
import { isConfigRef } from './ConfigRefs.js';
import { getFlows } from './normalizeGraph.js';
import { buildEndpointUri } from './UriBuilder.js';

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

function hasCycle(flow: FlowDefinition): boolean {
  const adjacency = new Map<string, string[]>();
  for (const node of flow.nodes) adjacency.set(node.id, []);
  for (const edge of flow.edges) {
    adjacency.get(edge.source)?.push(edge.target);
  }

  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map<string, number>(flow.nodes.map((n) => [n.id, WHITE]));

  function dfs(id: string): boolean {
    color.set(id, GRAY);
    for (const neighbor of adjacency.get(id) ?? []) {
      if (color.get(neighbor) === GRAY) return true;
      if (color.get(neighbor) === WHITE && dfs(neighbor)) return true;
    }
    color.set(id, BLACK);
    return false;
  }

  for (const node of flow.nodes) {
    if (color.get(node.id) === WHITE && dfs(node.id)) return true;
  }
  return false;
}

function validateFlow(
  flow: FlowDefinition,
  graph: FlowGraph,
  routeIds: Set<string>
): string[] {
  const errors: string[] = [];
  const prefix = `[${flow.name}] `;

  const sourceNodes = flow.nodes.filter(
    (n) => getBlock(n.blockType)?.category === BlockCategory.SOURCE
  );
  const destNodes = flow.nodes.filter(
    (n) => getBlock(n.blockType)?.category === BlockCategory.DESTINATION
  );

  if (flow.nodes.length > 0 && sourceNodes.length === 0) {
    errors.push(`${prefix}Flow must have at least one source block.`);
  }
  const hasCallFlow = flow.nodes.some((n) => n.blockType === 'call-flow-action');
  if (flow.nodes.length > 0 && destNodes.length === 0 && !hasCallFlow) {
    errors.push(`${prefix}Flow must have a destination or a Call flow block.`);
  }

  const incomingCount = new Map<string, number>();
  const outgoingCount = new Map<string, number>();

  for (const node of flow.nodes) {
    incomingCount.set(node.id, 0);
    outgoingCount.set(node.id, 0);
  }

  for (const edge of flow.edges) {
    outgoingCount.set(edge.source, (outgoingCount.get(edge.source) ?? 0) + 1);
    incomingCount.set(edge.target, (incomingCount.get(edge.target) ?? 0) + 1);
  }

  for (const node of flow.nodes) {
    const category = getBlock(node.blockType)?.category;
    const incoming = incomingCount.get(node.id) ?? 0;
    const outgoing = outgoingCount.get(node.id) ?? 0;

    if (category === BlockCategory.SOURCE && incoming > 0) {
      errors.push(`${prefix}Source block "${node.label}" must not have incoming connections.`);
    }
    if (category === BlockCategory.DESTINATION && outgoing > 0) {
      errors.push(`${prefix}Destination block "${node.label}" must not have outgoing connections.`);
    }
    if (category !== BlockCategory.SOURCE && incoming === 0) {
      errors.push(`${prefix}Block "${node.label}" has no incoming connection.`);
    }
    if (
      category !== BlockCategory.DESTINATION &&
      category !== BlockCategory.ACTION &&
      outgoing === 0
    ) {
      errors.push(`${prefix}Block "${node.label}" has no outgoing connection.`);
    }
    if (node.blockType === 'call-flow-action' && outgoing > 0) {
      errors.push(`${prefix}"${node.label}" (Call flow) must not have outgoing connections.`);
    }
  }

  if (hasCycle(flow)) {
    errors.push(`${prefix}Flow contains a cycle, which is not allowed.`);
  }

  return errors;
}

function validateFlowForYaml(
  flow: FlowDefinition,
  graph: FlowGraph,
  routeIds: Set<string>
): string[] {
  const errors = validateFlow(flow, graph, routeIds);
  const prefix = `[${flow.name}] `;

  for (const node of flow.nodes) {
    const block = getBlock(node.blockType);
    if (!block) {
      errors.push(`${prefix}Unknown block type "${node.blockType}" on "${node.label}".`);
      continue;
    }

    const props = resolveNodeProps(node.blockType, node.props);

    for (const schema of getWizardSteps(node.blockType)) {
      if (!schema.required) continue;
      if (!resolvePropValue(props, schema)) {
        errors.push(
          `${prefix}"${node.label}" (${block.label}): missing "${schema.label || schema.key}".`
        );
      }
    }

    if (node.blockType === 'call-flow-action') {
      const target = props['targetRouteId']?.trim();
      if (!target) {
        errors.push(`${prefix}"${node.label}" (Call flow): select a target flow.`);
      } else if (target === flow.routeId) {
        errors.push(`${prefix}"${node.label}" (Call flow): cannot call the same flow.`);
      } else if (!routeIds.has(target)) {
        errors.push(`${prefix}"${node.label}" (Call flow): unknown route "${target}".`);
      }
    }

    const needsUri =
      block.category === BlockCategory.SOURCE ||
      block.category === BlockCategory.DESTINATION ||
      block.camelUri !== '';
    if (needsUri && node.blockType !== 'call-flow-action') {
      const uri = buildEndpointUri(node.blockType, props);
      if (!uri) {
        errors.push(
          `${prefix}"${node.label}" (${block.label}): endpoint URI could not be built — check required settings.`
        );
      }
    }

    const eip = block.category === BlockCategory.ACTION;
    const expr = props['expression'];
    if (eip && node.blockType === 'filter-action' && !expr?.trim() && !isConfigRef(expr)) {
      errors.push(`${prefix}"${node.label}" (Filter): missing filter expression.`);
    }
    if (eip && node.blockType === 'transform-action' && !expr?.trim() && !isConfigRef(expr)) {
      errors.push(`${prefix}"${node.label}" (Transform): missing expression.`);
    }

    for (const [propKey, propVal] of Object.entries(node.props)) {
      if (isConfigRef(propVal)) {
        const cfgKey = propVal.slice('@config:'.length);
        if (!configKeyExists(graph.config, cfgKey)) {
          errors.push(
            `${prefix}"${node.label}": property "${propKey}" references unknown config key "${cfgKey}". Add it under Application properties.`
          );
        }
      }
    }
  }

  return errors;
}

export function validateForYamlExport(graph: FlowGraph): ValidationResult {
  const flows = getFlows(graph);
  const errors: string[] = [];

  const routeIdCounts = new Map<string, number>();
  for (const f of flows) {
    const rid = f.routeId?.trim() || '';
    if (rid) routeIdCounts.set(rid, (routeIdCounts.get(rid) ?? 0) + 1);
  }
  for (const [rid, count] of routeIdCounts) {
    if (count > 1) errors.push(`Duplicate route id "${rid}" across flows.`);
  }

  const routeIds = new Set(flows.map((f) => f.routeId).filter(Boolean));

  for (const flow of flows) {
    if (flow.nodes.length === 0) continue;
    errors.push(...validateFlowForYaml(flow, graph, routeIds));
  }

  if (flows.every((f) => f.nodes.length === 0)) {
    errors.push('Project has no blocks in any flow.');
  }

  return { valid: errors.length === 0, errors };
}

export function validate(graph: FlowGraph): ValidationResult {
  const flows = getFlows(graph);
  const errors: string[] = [];
  const routeIds = new Set(flows.map((f) => f.routeId).filter(Boolean));

  for (const flow of flows) {
    if (flow.nodes.length === 0) continue;
    errors.push(...validateFlow(flow, graph, routeIds));
  }

  if (flows.every((f) => f.nodes.length === 0)) {
    errors.push('Add at least one block to a flow.');
  }

  return { valid: errors.length === 0, errors };
}
